scheds/rust/scx_rusty/src/bpf/main.bpf.c

```diff
@@ -440,32 +505,225 @@ static void refresh_tune_params(void)
+	 * the producer or consumer is improved by a factor of x, it improves
+	 * the latency of the entire workload by:
+	 *
+	 *	S_lat(x) = 1 / ((1 - .25) + (.25 / x))
+	 *
+	 * Say that we improve wakeup latency by 2x for either task, the
+	 * latency improvement would be:
+	 *
+	 *	S_lat(2) = 1 / ((1 - .25) + (.25 / 2))
+	 *		 = 1 / (.75 + .125)
+	 *		 = 1 / .875
+	 *		~= 14.2%
+	 *
+	 * If we instead improve wakeup latency by 2x for all the n parallel
+	 * tasks in the latter 50% of the workload window, the improvement
+	 * would be:
+	 *
+	 *	S_lat(2) = 1 / ((1 - .5) + (.5 / 2))
+	 *		 = 1 / (.5 + .25)
+	 *		 = 1 / .75
+	 *		~= 33%
+	 *
+	 * This is also significant, but the returns are amortized across all
+	 * of those tasks. Thus, by giving a latency boost to the producer /
+	 * consumer tasks, we optimize for the case of scheduling tasks that
+	 * are on the critical path for serial workchains, and have a
+	 * disproporionate impact on the latency of a workload.
+	 *
+	 * We multiply the frequencies of wait_freq and waker_freq somewhat
+	 * arbitrarily, based on observed performance for audio and gaming
+	 * interactive workloads.
+	 */
+	waker_freq = max_clamp(taskc->waker_freq, DL_FREQ_FT_MAX);
+	blocked_freq = max_clamp(taskc->blocked_freq, DL_FREQ_FT_MAX);
+	freq_factor = blocked_freq * waker_freq * waker_freq;
+	freq_factor = scale_fair(freq_factor, p->scx.weight);
+
+	/*
+	 * The above frequencies roughly follow an exponential distribution, so
+	 * borrow the bpf_log2l() implementation from lavd to linearize it to a
+	 * boost priority.
+	 */
+	lat_prio = bpf_log2l(freq_factor + 1);
+	lat_prio = max_clamp(lat_prio, DL_MAX_LAT_PRIO);
+
+	/*
+	 * Next calculate a task's average runtime, and apply it to deadline
+	 * accordingly. A task with a large runtime is penalized from an
+	 * interactivity standpoint, for obvious reasons.
+	 *
+	 * As with waker and blocked frequencies above, this follows an
+	 * exponential distribution. We inversely scale to account for
+	 * empirical observations which seem to bring it roughly to the same
+	 * order of magnitude as the blocker and waker frequencies above.
+	 */
+	avg_run_raw = taskc->avg_runtime / DL_RUNTIME_SCALE;
+	avg_run_raw = max_clamp(avg_run_raw, DL_MAX_LATENCY_NS);
+	avg_run_raw = scale_fair(avg_run_raw, p->scx.weight);
+	avg_run = bpf_log2l(avg_run_raw + 1);
+
+	if (avg_run < lat_prio) {
+		lat_prio -= avg_run;
+	} else {
+		lat_prio = 0;
+	}
+
+	lat_scale = sched_prio_to_latency_weight(lat_prio);
+	lat_scale = max_clamp(lat_scale, LB_MAX_WEIGHT);
+
+	return scale_fair(taskc->avg_runtime, lat_scale);
+}
+
+static void clamp_task_vtime(struct task_struct *p, struct task_ctx *taskc, u64 enq_flags)
+{
+	u64 dom_vruntime, min_vruntime;
+	struct dom_ctx *domc;
+
+	if (!(domc = lookup_dom_ctx(taskc->dom_id)))
+		return;
+
+	dom_vruntime = dom_min_vruntime(domc);
+	min_vruntime = dom_vruntime - slice_ns;
+	if (vtime_before(p->scx.dsq_vtime, min_vruntime)) {
```

Comment:
Hmm... this if statement is not quite intuitive to me. Could you please share the idea behind this if statement?

Reply:
The idea here is that we're limiting the amount of vtime budget that a task can accumulate to be at most one slice. If we didn't clamp this here, then a task could sleep for e.g. 1 day, and then come back and have essentially infinite CPU capacity until it caught up hours later. I'll add a comment explaining as much.